Add e2e tests for ceph-csi(rbd/cephfs) volume

mkimuram opened the pull request:
**What this PR does / why we need it**:
Add e2e tests for ceph-csi(rbd/cephfs) volume.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #66388

**Special notes for your reviewer**:
/sig storage
@rootfs @aruneli @wongma7

Due to the issue discussing in https://github.com/ceph/ceph-csi/issues/48, cephfs test doesn't work well now.
I talked with @rootfs and decided to separate cephfs and rbd. So, this PR enables rbd only.
(Codes for cephfs exists in this PR, but is disabled by commenting out registering init function for it. If needed, I will delete the cephfs codes.)

@jsafrane @msau42 

I plan to apply the same refactoring as https://github.com/kubernetes/kubernetes/issues/66571 to csi tests, once it and this PR are merged.

**Release note**:
```release-note
NONE
```

mkimuram committed:
Add e2e tests for ceph-csi(rbd/cephfs) dynamic provision

Add e2e tests for ceph-csi(rbd/cephfs) dynamic provision.

fixes: #66388

wongma7 commented:
/ok-to-test

mkimuram commented:
FYI: 

> I plan to apply the same refactoring as #66571 to csi tests, once it and this PR are merged.

Once they are merged and apply the same refactoring to csi tests, they look like [this](https://github.com/mkimuram/kubernetes/commit/b93721645d07c1fe617637f32ffabedfa6d81fbb)
(I forget to commit after merge, so there are many differences, however, only the diff in csi_volumes.go and volume_provisioning.go will be the actual difference.)

I only tested with CSI rbd and Non-CSI NFS, but it seems to work well.

mkimuram commented:
/test pull-kubernetes-integration

mkimuram commented:
/test pull-kubernetes-integration

rootfs commented on test/e2e/storage/csi_objects.go:
```diff
@@ -70,14 +69,30 @@ func csiDriverRegistrarClusterRole(
 	superuserClientset, err := clientset.NewForConfig(rc)
 	framework.ExpectNoError(err, "Failed to create superuser clientset: %v", err)
-	By("Creating the CSI driver registrar cluster role")
+	By(fmt.Sprintf("Creating the %s cluster role", role.ObjectMeta.Name))
```
nit: role.GetName() instead of role.ObjectMeta.Name

mkimuram replied:
Fixed as suggested.

rootfs commented on test/e2e/storage/csi_objects.go:
```diff
@@ -70,14 +69,30 @@ func csiDriverRegistrarClusterRole(
+
+	ret, err := clusterRoleClient.Create(role)
+	if err != nil {
+		if apierrs.IsAlreadyExists(err) {
```
i think this should be `if ! apierrs...`

rootfs commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -48,6 +49,8 @@ var csiTestDrivers = map[string]func(f *framework.Framework, config framework.Vo
 	"hostPath": initCSIHostpath,
 	// Feature tag to skip test in CI, pending fix of #62237
 	"[Feature: GCE PD CSI Plugin] gcePD": initCSIgcePD,
+	"[Feature:Volumes] rbd":              initCSIrbd,
```
"[Feature:Volumes] rbd csi driver" (so we can just run csi driver explicitly in e2e test)

mkimuram replied:
Fixed like "csi xxx" instead "xxx csi driver", because the test names tend to be long.

rootfs commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -223,3 +227,144 @@ func (g *gcePDCSIDriver) cleanupCSIDriver() {
+		claimSize:          "1Gi",
+		expectedSize:       "1Gi",
+		nodeName:           node.Name,
+		skipWriteReadCheck: true,
```
why not test r/w?

mkimuram replied:
If skipWriteReadCheck is false, both rbd and cephfs tests fail in second 
[```runInPodWithVolume```](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/volume_provisioning.go#L153) run. I will file another issues for csi cephfs and rbd.

Until the issues are fixed, is it O.K. to merge this PR with ```skipWriteReadCheck=true```?

rootfs replied:
thanks!

rootfs commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -223,3 +227,144 @@ func (g *gcePDCSIDriver) cleanupCSIDriver() {
+		claimSize:          "1Gi",
+		expectedSize:       "1Gi",
+		nodeName:           node.Name,
+		skipWriteReadCheck: true,
```
why skip?

mkimuram committed:
Fix to reflect review comments

- Fix how to access role name, role.GetName() instead of role.ObjectMeta.Name
- Fix driver name to explicitly show that they are csi drivers

mkimuram commented:
@rootfs 
> If skipWriteReadCheck is false, both rbd and cephfs tests fails ```runInPodWithVolume```. 

I filed two separate issues for rbd and cephfs. Please check them.
(I'm still not sure if I'm doing something wrong in my e2e test code side, though.)

- [[csi-rbd] Mount volume fails when PVC is reused to second pod after first pod is deleted](https://github.com/ceph/ceph-csi/issues/67)
- [[csi-cephfs] Mount volume fails with failed to find stage secret](https://github.com/ceph/ceph-csi/issues/68)

gman0 commented on test/e2e/testing-manifests/storage-csi/cephfs/node_ds.yaml:
```diff
@@ -0,0 +1,84 @@
+              add: ["SYS_ADMIN"]
+            allowPrivilegeEscalation: true
+          imagePullPolicy: IfNotPresent
+          image: quay.io/cephcsi/cephfsplugin:v0.3.0
```
Sorry for responding late! You're using csi-cephfs 0.3.0 here, but the external-provisioner, external-attacher and driver-registrar are at 0.2.0 - you should have all those at `v0.3.0` (AFAIK external-provisioner populates node stage secrets reference from 0.3.0+).

gman0 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -223,3 +227,144 @@ func (g *gcePDCSIDriver) cleanupCSIDriver() {
+			"monitors":        c.serverIP,
+			"provisionVolume": "true",
+			"pool":            "cephfs_data",
+			"csiProvisionerSecretName":      c.secret.Name,
```
This changed in with 0.3.0 (along with refurbished docs), please have a look at https://github.com/ceph/ceph-csi/blob/master/docs/deploy-cephfs.md#configuration

```
...
"csiProvisionerSecretName": c.secret.Name,
"csiProvisionerSecretNamespace": c.config.Namespace,
"csiNodeStageSecretName": c.secret.Name,
"csiNodeStageSecretNamespace": c.config.Namespace,
...
```

mkimuram committed:
Update csi driver version

mkimuram commented:
@ixdy 

Thank you for caring about this PR.

This e2e test case won't be tested directly by the CI system, because  [Feature: Volume] tag is added so that it won't be tested in environments that doesn't have tool to mount ceph/rbd volumes.
Instead, it will need to be executed manually.
(I believe that @jsafrane is doing another work to make this kind of test work by running daemon set that have tools inside. So once it will complete, it should be automatically tested.)

Therefore, even if all the tests above succeed, it doesn't mean that above issue is resolved.
(I'm not sure if we can merge tests when it is harmless to CI even though the tests fails, though.)

mkimuram committed:
Add configurations to handle kubelet-plugin-watcher

mkimuram commented:
Apply the same change to https://github.com/ceph/ceph-csi/issues/73 .

pohly commented:
I am working on changing the CSI tests such that they create all relevant entities (driver deployment, storage class) from .yaml files. That will make it easier to ensure that tests use exactly the same setup as real usages. Just FYI, this PR can also be merged as-is and be rewritten later.

However, I am wondering: what is the motivation for having a test for ceph-csi in Kubernetes itself? Are there Kubernetes CI jobs were the test will be enabled? Or is the motivation merely that this is currently the easiest way of creating E2E tests for ceph-csi?

msau42 commented:
Once https://github.com/kubernetes/kubernetes/pull/68025 goes in, I'm actually thinking about removing the GCE PD CSI driver tests from kubernetes/kubernetes, and have them run off of the csi driver repo instead.  This will make it easier to test canary and stable versions of the kubernetes, csi driver, csi sidecar containers.  And driver setup can then be delegated outside of the e2e tests or done in the BeforeSuite, which lets us parallelize the tests better.

pohly commented:
Michelle Au <notifications@github.com> writes:
> Once https://github.com/kubernetes/kubernetes/pull/68025
> goes in, I'm actually thinking about removing the GCE PD CSI driver tests from
> kubernetes/kubernetes, and have them run off of the csi driver repo
> instead.

kubernetes-csi/drivers might not be the right place, considering that
it's supposed to get split up. I was thinking about a separate csi-e2e
repo which contains just the E2E testing and CI work that is required to
take the individual components and produce a "Kubernetes CSI"
release. Whether that includes third-party drivers like ceph-csi would
have to be discussed.

> This will make it easier to test canary and stable versions
> of the kubernetes, csi driver, csi sidecar containers.  And driver
> setup can then be delegated outside of the e2e tests or done in the
> BeforeSuite, which lets us parallelize the tests better.

See my comment on https://github.com/kubernetes/kubernetes/pull/68025: I
have code which can do this setup such that tests can run in parallel.

mkimuram commented:
> I'm actually thinking about removing the GCE PD CSI driver tests from kubernetes/kubernetes, and have them run off of the csi driver repo instead.

I think that the purpose of the test would be different between tests in k8s repo and tests in each csi driver's repo.

  1. To confirm that all csi drivers of particular versions work with the latest k8s(To confirm that the change in k8s doesn't break anything) ... tested in k8s e2e test
  2. To confirm that each latest csi driver works with a particular version of k8s(To confirm that the change in the driver doesn't break anything)  ...  tested in each csi driver's e2e test

So, I guess that it is worth keeping them in both repos. I can't say exact which set of tests should be covered in which repo instantly, though.
For example, if we only do csi test for hostpath in k8s repo, feature tests that are not valid for hostpath, such as block volume, volume topology, and volume expansion, won't be tested properly. Therefore, we will still need at least reference drivers that cover all features in k8s repo and these kind of features should be tested there.  (Making all csi drivers tested for all set of tests would still be a candidate until the number of the test explodes.)

> I am working on changing the CSI tests such that they create all relevant entities (driver deployment, storage class) from .yaml files. That will make it easier to ensure that tests use exactly the same setup as real usages.

It would be very nice to improve the way to deploy csi drivers. It is out of the scope of #68025 and is missing in current e2e code.

msau42 commented:
I opened up https://github.com/kubernetes/kubernetes/issues/69819 with my ideas on how we should do Kubernetes e2e testing of third-party csi drivers.

mkimuram commented:
@msau42 

Thank you for summarizing the discussion and your idea. Let's discuss general e2e testing topic in #69819.
